### PR TITLE
Strain optimization

### PR DESCRIPTION
**Pull request by ejpaul**
This PR is based on work by @phuslage for the optimization of coil curves and the HTS winding angle to reduce torsional and hard-way bending strain as defined in : 

    Paz Soldan, "Non-planar coil winding angle optimization for compatibility with 
    non-insulated high-temperature superconducting magnets", Journal of Plasma Physics 
    86 (2020), doi:10.1017/S0022377820001208. 

The orientation of the HTS tape normal is defined by the FramedCurve class, which is inherited by FramedCurveFrenet or FramedCurveCentroid. The FramedCurve defines a curve and an orthonormal frame, where the tangent vector is aligned with the curve tangent. The frame can be rotated by an angle alpha with respect to the Frenet or centroid reference frames. The FramedCurve class evaluates geometric properties of the frame, such as the torsion and binormal curvature. This class is also used to set up a multi-filament coil using the CurveFilament class. 

The strains are evaluated in the StrainOpt class, which is defined by a FramedCurve and a tape width. The objective functions LPTorsionalStrainPenalty and LPBinormalCurvatureStrainPenalty can be used for optimization with respect to the coil or rotation dofs. 

An example of optimization with respect to the rotation dofs is provided in strain_optimization_simple.py

### CODE REVIEW

File: src/simsopt/field/self_field.py
Comment (phuslage) [inline src/simsopt/field/self_field.py:1]: This is an accidental commit which was meant for the coil_forces so we should delete it.

File: examples/2_Intermediate/strain_optimization_simple.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+The orientation of the tape is defined wrt the Frenet-Serret Frame
+"""
+
+import numpy as np
+import os
+import numpy as np
+from scipy.optimize import minimize
+from simsopt.geo import StrainOpt, LPTorsionalStrainPenalty, LPBinormalCurvatureStrainPenalty
+from simsopt.geo import FrameRotation, FramedCurveFrenet, CurveXYZFourier
+from simsopt.field import load_coils_from_makegrid_file
```
Comment (landreman) [inline examples/2_Intermediate/strain_optimization_simple.py]: Remove unnecessary import

File: CMakeLists.txt [hunk 2]
```diff
@@ -51,7 +51,7 @@ else()
     unset(COMPILER_SUPPORTS_MARCH_NATIVE CACHE)
     CHECK_CXX_COMPILER_FLAG(-march=native COMPILER_SUPPORTS_MARCH_NATIVE)
     if(COMPILER_SUPPORTS_MARCH_NATIVE)
-    	set(CMAKE_CXX_FLAGS "-O3 -march=native -mfma -ffp-contract=fast")
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply CMakeLists.txt]: I reverted this change.

File: examples/2_Intermediate/strain_optimization_simple.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+"""
+This script performs an optimization of the HTS tape winding angle 
+with respect to binormal curvature and torsional strain cost functions. 
+The orientation of the tape is defined wrt the Frenet-Serret Frame
+"""
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply examples/2_Intermediate/strain_optimization_simple.py]: Done

File: examples/2_Intermediate/strain_optimization_simple.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+import numpy as np
+from scipy.optimize import minimize
+from simsopt.geo import StrainOpt, LPTorsionalStrainPenalty, LPBinormalCurvatureStrainPenalty
+from simsopt.geo import FrameRotation, FramedCurveFrenet, CurveXYZFourier
+from simsopt.field import load_coils_from_makegrid_file
+from simsopt.configs import get_hsx_data
+import matplotlib.pyplot as plt
+
+ci = "CI" in os.environ and os.environ['CI'].lower() in ['1', 'true']
+MAXITER = 50 if ci else 400
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply examples/2_Intermediate/strain_optimization_simple.py]: Done

File: examples/3_Advanced/strain_simple.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+"""
```
[the remark this answers is not in the document]
Comment (ejpaul) [reply examples/3_Advanced/strain_simple.py]: I didn't mean to include this example. It is deleted.

**Review by phuslage — APPROVED** [review]

**Comment by landreman** [comment]
Is this PR ready for another review? If so can you merge master in and resolve merge conflicts?

**Comment by ejpaul** [comment]
> Is this PR ready for another review? If so can you merge master in and resolve merge conflicts?

@landreman Yes, it is ready. I have merged in master.

**Review by landreman — APPROVED** [review]
Looks good. Since `StrainOpt` was not intended to be used as an objective function in optimization, the `Opt` in the name might be confusing to users. So I took the liberty of renaming it to `CoilStrain`, and adding to the docs to clarify. Hope that's ok. Feel free to merge whenever you're ready.

**Comment by ejpaul** [comment]
> Looks good. Since `StrainOpt` was not intended to be used as an objective function in optimization, the `Opt` in the name might be confusing to users. So I took the liberty of renaming it to `CoilStrain`, and adding to the docs to clarify. Hope that's ok. Feel free to merge whenever you're ready.

Thanks, Matt!

**Pull request merged**